<a href="https://colab.research.google.com/github/CaioSepulveda/Analysis_DataSUS_2024/blob/main/DataSUS2024.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [26]:
import pandas as pd
dados = pd.read_csv('/content/drive/MyDrive/Analise_de_Dados/Datasets/DATASUS/2024.csv', encoding='latin-1', sep=';', on_bad_lines='skip')
dados.head()

Nome Instituição    CNPJ Instituição Município Instituição  UF  \
0  FUNDO MUNICIPAL DE SAUDE  11.120.699/0001-40                MURICI  AL   
1  FUNDO MUNICIPAL DE SAUDE  11.120.699/0001-40                MURICI  AL   
2  FUNDO MUNICIPAL DE SAUDE  11.120.699/0001-40                MURICI  AL   
3  FUNDO MUNICIPAL DE SAUDE  11.120.699/0001-40                MURICI  AL   
4  FUNDO MUNICIPAL DE SAUDE  11.120.699/0001-40                MURICI  AL   

   Compra  Inserção Modalidade da Compra  Código BR  \
0   45292   45513.0               Pregão     365246   
1   45292   45513.0               Pregão     271620   
2   45292   45513.0               Pregão     328530   
3   45292   45513.0               Pregão     448808   
4   45292   45513.0               Pregão     480364   

                                    Descrição CATMAT Unidade Fornecimento  \
0  EXTRATO, TIPO:EXTRATO GLICÓLICO, NOME COMUM:BA...                LITRO   
1                           OLANZAPINA, DOSAGEM:5 MG           COMPRIMIDO   
2            VALPROATO DE SÓDIO, CONCENTRAÇÃO:500 MG           COMPRIMIDO   
3  NORETISTERONA, CONCENTRAÇAO:0,35 MG, CARACTERÍ...              BLISTER   
4  EQUIPO ESPECIAL, APLICAÇÃO:P/ INJEÇÃO GÁS CARB...              UNIDADE   

  Genérico  ANVISA     CNPJ Fornecedor  \
0        N     NaN  00.236.193/0001-84   
1        N     NaN  00.236.193/0001-84   
2        N     NaN  00.236.193/0001-84   
3        N     NaN  00.236.193/0001-84   
4        N     NaN  00.236.193/0001-84   

                                        Fornecedor     CNPJ Fabricante  \
0  CIRURGICA RECIFE COMERCIO E REPRESENTACOES LTDA  33.109.400/0000-00   
1  CIRURGICA RECIFE COMERCIO E REPRESENTACOES LTDA  33.109.400/0000-00   
2  CIRURGICA RECIFE COMERCIO E REPRESENTACOES LTDA  33.109.400/0000-00   
3  CIRURGICA RECIFE COMERCIO E REPRESENTACOES LTDA  33.109.400/0000-00   
4  CIRURGICA RECIFE COMERCIO E REPRESENTACOES LTDA  33.109.400/0000-00   

                                          Fabricante  Qtd Itens Comprados  \
0  CASA GRANADO LABORATORIOS FARMACIAS E DROGARIA...                  360   
1  CASA GRANADO LABORATORIOS FARMACIAS E DROGARIA...                 3060   
2  CASA GRANADO LABORATORIOS FARMACIAS E DROGARIA...                24600   
3  CASA GRANADO LABORATORIOS FARMACIAS E DROGARIA...                 4200   
4  CASA GRANADO LABORATORIOS FARMACIAS E DROGARIA...                 7200   

  Preço Unitário Preço Total  
0           2,66       957,6  
1            1,1        3366  
2           1,16       28536  
3           0,37        1554  
4           6,74       48528

In [27]:
# This code loads the data from the CSV file into a pandas DataFrame.
# I added 'encoding='latin-1'' because the initial attempt to read the file with default UTF-8 encoding resulted in a UnicodeDecodeError.
# The error indicated that the file was likely not encoded in UTF-8, and 'latin-1' is a common alternative encoding for files that contain special characters or were created on different systems.
# I also added 'sep=';'' because a ParserError suggested that the delimiter was not the default comma. The error message implied an inconsistent number of fields, which often happens when the wrong delimiter is used.
# Finally, I added 'on_bad_lines='skip'' to handle any lines that still might cause parsing errors, allowing the rest of the data to be loaded.

import pandas as pd
dados = pd.read_csv('/content/drive/MyDrive/Analise_de_Dados/Datasets/DATASUS/2024.csv', encoding='latin-1', sep=';', on_bad_lines='skip')
dados.head()

Nome Instituição    CNPJ Instituição Município Instituição  UF  \
0  FUNDO MUNICIPAL DE SAUDE  11.120.699/0001-40                MURICI  AL   
1  FUNDO MUNICIPAL DE SAUDE  11.120.699/0001-40                MURICI  AL   
2  FUNDO MUNICIPAL DE SAUDE  11.120.699/0001-40                MURICI  AL   
3  FUNDO MUNICIPAL DE SAUDE  11.120.699/0001-40                MURICI  AL   
4  FUNDO MUNICIPAL DE SAUDE  11.120.699/0001-40                MURICI  AL   

   Compra  Inserção Modalidade da Compra  Código BR  \
0   45292   45513.0               Pregão     365246   
1   45292   45513.0               Pregão     271620   
2   45292   45513.0               Pregão     328530   
3   45292   45513.0               Pregão     448808   
4   45292   45513.0               Pregão     480364   

                                    Descrição CATMAT Unidade Fornecimento  \
0  EXTRATO, TIPO:EXTRATO GLICÓLICO, NOME COMUM:BA...                LITRO   
1                           OLANZAPINA, DOSAGEM:5 MG           COMPRIMIDO   
2            VALPROATO DE SÓDIO, CONCENTRAÇÃO:500 MG           COMPRIMIDO   
3  NORETISTERONA, CONCENTRAÇAO:0,35 MG, CARACTERÍ...              BLISTER   
4  EQUIPO ESPECIAL, APLICAÇÃO:P/ INJEÇÃO GÁS CARB...              UNIDADE   

  Genérico  ANVISA     CNPJ Fornecedor  \
0        N     NaN  00.236.193/0001-84   
1        N     NaN  00.236.193/0001-84   
2        N     NaN  00.236.193/0001-84   
3        N     NaN  00.236.193/0001-84   
4        N     NaN  00.236.193/0001-84   

                                        Fornecedor     CNPJ Fabricante  \
0  CIRURGICA RECIFE COMERCIO E REPRESENTACOES LTDA  33.109.400/0000-00   
1  CIRURGICA RECIFE COMERCIO E REPRESENTACOES LTDA  33.109.400/0000-00   
2  CIRURGICA RECIFE COMERCIO E REPRESENTACOES LTDA  33.109.400/0000-00   
3  CIRURGICA RECIFE COMERCIO E REPRESENTACOES LTDA  33.109.400/0000-00   
4  CIRURGICA RECIFE COMERCIO E REPRESENTACOES LTDA  33.109.400/0000-00   

                                          Fabricante  Qtd Itens Comprados  \
0  CASA GRANADO LABORATORIOS FARMACIAS E DROGARIA...                  360   
1  CASA GRANADO LABORATORIOS FARMACIAS E DROGARIA...                 3060   
2  CASA GRANADO LABORATORIOS FARMACIAS E DROGARIA...                24600   
3  CASA GRANADO LABORATORIOS FARMACIAS E DROGARIA...                 4200   
4  CASA GRANADO LABORATORIOS FARMACIAS E DROGARIA...                 7200   

  Preço Unitário Preço Total  
0           2,66       957,6  
1            1,1        3366  
2           1,16       28536  
3           0,37        1554  
4           6,74       48528

In [28]:
dados.isna().any()

Nome Instituição          True
CNPJ Instituição         False
Município Instituição    False
UF                       False
Compra                   False
Inserção                  True
Modalidade da Compra     False
Código BR                False
Descrição CATMAT         False
Unidade Fornecimento      True
Genérico                 False
ANVISA                    True
CNPJ Fornecedor          False
Fornecedor               False
CNPJ Fabricante          False
Fabricante               False
Qtd Itens Comprados      False
Preço Unitário           False
Preço Total              False
dtype: bool

# Task
Clean the provided dataset.

## Handle missing values

### Subtask:
Address the missing values in the columns 'Nome Instituição', 'Inserção', 'Unidade Fornecimento', and 'ANVISA'. This could involve dropping rows or columns, or imputing missing values based on the nature of the data in each column.


**Reasoning**:
Calculate and display the percentage of missing values for the specified columns to inform the decision on how to handle them.



In [29]:
missing_percentage = dados[['Nome Instituição', 'Inserção', 'Unidade Fornecimento', 'ANVISA']].isnull().mean() * 100
display(missing_percentage)

Nome Instituição         0.044652
Inserção                14.272377
Unidade Fornecimento     0.060889
ANVISA                  61.891618
dtype: float64

**Reasoning**:
Based on the missing value percentages, 'ANVISA' has a high percentage of missing values (over 60%), making imputation or dropping rows less effective. 'Nome Instituição' and 'Unidade Fornecimento' have very low percentages, so dropping rows with missing values in these columns is a reasonable approach to retain most of the data. 'Inserção' has a moderate percentage (around 14%), and since its date-related, imputing with the median is a suitable strategy to preserve the distribution.



In [30]:
# Drop the 'ANVISA' column due to a high percentage of missing values
dados = dados.drop('ANVISA', axis=1)

# Drop rows with missing values in 'Nome Instituição' and 'Unidade Fornecimento'
dados = dados.dropna(subset=['Nome Instituição', 'Unidade Fornecimento'])

# Impute missing values in 'Inserção' with the median
if pd.api.types.is_numeric_dtype(dados['Inserção']):
    median_insercao = dados['Inserção'].median()
    dados['Inserção'].fillna(median_insercao, inplace=True)
else:
    # If 'Inserção' is not numeric, impute with mode or a placeholder
    # Check data type to determine appropriate imputation
    mode_insercao = dados['Inserção'].mode()[0] if not dados['Inserção'].mode().empty else 'Unknown'
    dados['Inserção'].fillna(mode_insercao, inplace=True)

# Verify that missing values have been addressed in the target columns
print("Missing values after cleaning:")
display(dados[['Nome Instituição', 'Inserção', 'Unidade Fornecimento']].isnull().sum())

Missing values after cleaning:


<ipython-input-30-669331661>:10: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  dados['Inserção'].fillna(median_insercao, inplace=True)


Nome Instituição        0
Inserção                0
Unidade Fornecimento    0
dtype: int64

## Check data types

### Subtask:
Verify that the data types of each column are appropriate for analysis and convert them if necessary (e.g., convert 'Preço Unitário' and 'Preço Total' to numeric types).


**Reasoning**:
Display the data types of all columns to identify columns that need type conversion.



In [31]:
display(dados.dtypes)

Nome Instituição          object
CNPJ Instituição          object
Município Instituição     object
UF                        object
Compra                     int64
Inserção                 float64
Modalidade da Compra      object
Código BR                  int64
Descrição CATMAT          object
Unidade Fornecimento      object
Genérico                  object
CNPJ Fornecedor           object
Fornecedor                object
CNPJ Fabricante           object
Fabricante                object
Qtd Itens Comprados        int64
Preço Unitário            object
Preço Total               object
dtype: object

**Reasoning**:
Convert 'Preço Unitário' and 'Preço Total' columns to numeric by removing commas and converting to float, coercing errors to NaN. Then display the data types again to confirm the conversion.



In [32]:
# Convert 'Preço Unitário' to numeric
dados['Preço Unitário'] = dados['Preço Unitário'].astype(str).str.replace(',', '.', regex=False)
dados['Preço Unitário'] = pd.to_numeric(dados['Preço Unitário'], errors='coerce')

# Convert 'Preço Total' to numeric
dados['Preço Total'] = dados['Preço Total'].astype(str).str.replace(',', '.', regex=False)
dados['Preço Total'] = pd.to_numeric(dados['Preço Total'], errors='coerce')

# Display data types to confirm conversion
display(dados.dtypes)

Nome Instituição          object
CNPJ Instituição          object
Município Instituição     object
UF                        object
Compra                     int64
Inserção                 float64
Modalidade da Compra      object
Código BR                  int64
Descrição CATMAT          object
Unidade Fornecimento      object
Genérico                  object
CNPJ Fornecedor           object
Fornecedor                object
CNPJ Fabricante           object
Fabricante                object
Qtd Itens Comprados        int64
Preço Unitário           float64
Preço Total              float64
dtype: object

### Datetime conversion

The code cell below converts the 'Compra' and 'Inserção' columns to datetime objects.

- `pd.to_datetime()` is used for the conversion.
- `origin='1899-12-30'` is specified because the original numeric values appear to be OLE Automation dates, which use this date as their origin.
- `unit='D'` indicates that the numeric values represent the number of days since the origin.
- `errors='coerce'` is used for the 'Inserção' column to turn any values that cannot be parsed into datetime objects into `NaT` (Not a Time).

In [33]:
print("Data types:")
display(dados[['Compra', 'Inserção']].dtypes)

print("\nSample values from 'Compra':")
display(dados['Compra'].head())

print("\nSample values from 'Inserção':")
display(dados['Inserção'].head())

Data types:


Compra        int64
Inserção    float64
dtype: object


Sample values from 'Compra':


0    45292
1    45292
2    45292
3    45292
4    45292
Name: Compra, dtype: int64


Sample values from 'Inserção':


0    45513.0
1    45513.0
2    45513.0
3    45513.0
4    45513.0
Name: Inserção, dtype: float64

In [34]:
# Convert 'Compra' and 'Inserção' to datetime objects
dados['Compra'] = pd.to_datetime(dados['Compra'], origin='1899-12-30', unit='D')
dados['Inserção'] = pd.to_datetime(dados['Inserção'], origin='1899-12-30', unit='D', errors='coerce')

print("Data types after conversion:")
display(dados[['Compra', 'Inserção']].dtypes)

print("\nSample values from 'Compra' after conversion:")
display(dados['Compra'].head())

print("\nSample values from 'Inserção' after conversion:")
display(dados['Inserção'].head())

Data types after conversion:


Compra      datetime64[ns]
Inserção    datetime64[ns]
dtype: object


Sample values from 'Compra' after conversion:


0   2024-01-01
1   2024-01-01
2   2024-01-01
3   2024-01-01
4   2024-01-01
Name: Compra, dtype: datetime64[ns]


Sample values from 'Inserção' after conversion:


0   2024-08-09
1   2024-08-09
2   2024-08-09
3   2024-08-09
4   2024-08-09
Name: Inserção, dtype: datetime64[ns]

## Validate data ranges and formats

### Subtask:
Check if numerical data is within expected ranges and if categorical data follows a consistent format.


**Reasoning**:
Calculate and display descriptive statistics for the numerical columns to check for values outside a plausible range.



In [35]:
numerical_cols = ['Compra', 'Inserção', 'Qtd Itens Comprados', 'Preço Unitário', 'Preço Total']
display(dados[numerical_cols].describe())

Compra                       Inserção  \
count                          24609                          24609   
mean   2024-05-15 00:22:24.678776064  2024-09-15 07:44:15.565037056   
min              2024-01-01 00:00:00            2021-01-04 00:00:00   
25%              2024-03-13 00:00:00            2024-08-27 00:00:00   
50%              2024-05-28 00:00:00            2024-09-18 00:00:00   
75%              2024-07-10 00:00:00            2024-10-10 00:00:00   
max              2024-10-31 00:00:00            2025-02-12 00:00:00   
std                              NaN                            NaN   

       Qtd Itens Comprados  Preço Unitário   Preço Total  
count         2.460900e+04    2.460900e+04  2.460900e+04  
mean          6.114819e+04    2.482062e+02  8.598906e+04  
min           1.000000e+00    7.000000e-03  2.500000e+00  
25%           3.000000e+02    3.270000e-01  6.400000e+02  
50%           2.000000e+03    1.534000e+00  2.760000e+03  
75%           1.000000e+04    6.000000e+00  1.096200e+04  
max           1.454131e+08    2.038000e+06  9.592000e+08  
std           1.256112e+06    1.576335e+04  6.170786e+06

## Summary:

### Data Analysis Key Findings

*   The initial dataset had missing values in 'Nome Instituição' (0.04%), 'Inserção' (14.27%), 'Unidade Fornecimento' (0.06%), and 'ANVISA' (61.89%). The 'ANVISA' column was subsequently dropped due to the high percentage of missing values.
*   The 'Preço Unitário' and 'Preço Total' columns were successfully converted from object type to `float64`.
*   Numerical columns ('Compra', 'Inserção', 'Qtd Itens Comprados', 'Preço Unitário', 'Preço Total') show a wide range of values.


In [36]:
# Save the cleaned dataset to a new CSV file
dados.to_csv('/content/drive/MyDrive/Analise_de_Dados/Datasets/DATASUS/2024_clean.csv', index=False)

print("Cleaned dataset saved successfully as '2024_clean.csv'")

Cleaned dataset saved successfully as '2024_clean.csv'
